Navbar buttons Version1.0

Eeyore6 opened the pull request:
Add buttons(Home, Dashboard, Board and Profile). When a user is not logged in, the buttons will be hidden, and only the login/signup button will be there. Once the user is logged in, the page buttons will appear. Any log-out action will lead to going back to the home page.
![image](https://github.com/mursal-ashraf/bankan/assets/100513480/6c4a597e-c969-4f71-9693-6c3712d14f63)
![image](https://github.com/mursal-ashraf/bankan/assets/100513480/1760249c-1070-416a-9a9d-450de9a15d6a)

Eeyore6 committed:
Navbar buttons Version1.0

Add buttons(Home, Dashboard, Board and Profile). When a user is not logged in, the buttons will be hidden, and only the login/signup button will be there. Once the user is logged in, the page buttons will appear. Any log-out action will lead to going back to the home page.

mursal-ashraf commented on frontend/src/hooks/useLogOut.ts:
```diff
@@ -0,0 +1,31 @@
+  const performLogOut = useCallback<LogOutHookFunction>(
+    async () => {
+      setResult({ ...result, isLoading: true });
+      await client.auth.signOut();
```
The sign out function does have a chance to return error, so just make changes to take that into account.
```const { error } = await client.auth.signOut();```
```setResult({...result, isLoading: false, error})```

Eeyore6 replied:
Done

mursal-ashraf commented on frontend/src/hooks/useLogin.ts:
```diff
@@ -31,9 +31,19 @@ export const useLogin: LoginHook = () => {
       const { data, error } = await client.auth.signInWithPassword(credentials);
       setResult({ ...result, data, error });
+      if (data && !error) {
+        window.location.reload();
```
What's the reload for here?

Eeyore6 replied:
When the Login action is done and there is no error, it will reload pages and show the buttons. If not reload, then you have to refresh to update the navbar buttons

mursal-ashraf replied:
Try changing line 45 ish in AuthModel.tsx, it shouldn't be `!!error` rather it should be `!error`, as we should only perform the submit function when error is nil.

mursal-ashraf replied:
It's better to leave one overall, generic purpose in hooks. Something like this is reactive to what the hook does, better handled in externally rather than within a hook.

Eeyore6 replied:
Yeah I think so, I will try after a successful click then reload inside the click event

mursal-ashraf commented on frontend/src/common/navbar/index.tsx:
```diff
@@ -1,111 +1,82 @@
+  const handleLogOut = async (event: React.MouseEvent) => {
+    event.preventDefault(); // prevent default anchor behavior
+    await performLogOut();
+    window.location.href = '/'; // redirect to home after logout
```
Would prefer to use the `useNavigate` hook rather than the window variable. There are examples in other current PRs. It's a hook from the `react-router-dom` we are using to handle routing/navigation.

mursal-ashraf replied:
Also, I reckon you wouldn't need the comments, your code is quite self documenting here.

Eeyore6 replied:
Will remove the comments, tried to use the Navigate but idk why it always be a blank page

mursal-ashraf replied:
> Will remove the comments, tried to use the Navigate but idk why it always be a blank page

could you put a screenshot of what you were trying?

Eeyore6 replied:
The 19th line is used be like :
![image](https://github.com/mursal-ashraf/bankan/assets/100513480/7bb7bf68-3b09-4197-8d1d-4eb21f368631)

mursal-ashraf replied:
can you also attach how you define `navigate`, thx.

Eeyore6 replied:
![image](https://github.com/mursal-ashraf/bankan/assets/100513480/ce1f73e1-dd38-4f98-8903-8d6a16ee734d)

mursal-ashraf commented on frontend/src/hooks/useLogin.ts:
```diff
@@ -31,9 +31,19 @@ export const useLogin: LoginHook = () => {
       setResult({ ...result, isLoading: false });
       const { data, error } = await client.auth.signInWithPassword(credentials);
       setResult({ ...result, data, error });
```
Can you also fix one thing we forget when adding these hooks? set `isLoading` to true before we call supabase and set it to false after we get the result. Same for the `useSignUp` hook. That way I can close the PR I made.

Eeyore6 replied:
Done

mursal-ashraf commented on frontend/src/common/navbar/index.tsx:
```diff
@@ -1,111 +1,82 @@
+          {/* If the user is logged in, show dashboard, board, and profile links */}
+          {isLoggedIn && (
+            <>
+              <a href="/" style={{ textDecoration: 'none' }}>
```
Also, rather than hardcoding the hrefs, it's better to use the `Routes` enum (in AppRouter from memory).
That way we only define routes in one place, and in the future if we decided to modify them, we'll only have to modify it in one place rather than all over the codebase. Means less work for us :)

mursal-ashraf commented on frontend/src/common/navbar/index.tsx:
```diff
@@ -1,111 +1,82 @@
+                </Button>
+              </a>
+              <a href="/dashboard" style={{ textDecoration: 'none' }}>
+                <Button color="inherit">
```
Just another way to write the code could be:

import {useNavigate} from 'react-router-dom';
const navigateTo = useNavigate();

....
.....
....
<Button color="inherit" onClick={() => navigateTo(Routes.Profile)} style={{textDecoration: 'none'}}>
    Profile
</Button>

Eeyore6 replied:
I used the Navigate Function but it didnt work, instead it will become a empty page, thats why I use the hardcode.
And route works, I can replace the href with the route part.

mursal-ashraf replied:
> I used the Navigate Function but it didnt work, instead it will become a empty page, thats why I use the hardcode. And route works, I can replace the href with the route part.

can you show an example of what you were trying?

mursal-ashraf replied:
a screenshot would be best

Eeyore6 replied:
import { useNavigate } from 'react-router-dom';

interface NavbarProps {
  navItems: string[];
}

const NavBar = ({ navItems }: NavbarProps) => {
  const navigate = useNavigate();
  
  ......
  
  
           <Button color="inherit" onClick={() => navigator("/dashboard")}>
            Dashboard
          </Button>
          <Button color="inherit" onClick={() => navigator("/board")}>
            Board
          </Button>
          <Button color="inherit" onClick={() => navigator("/profile")}>
            Profile
          </Button>

Eeyore6 replied:
When I do the click it will nagivate to the path but not working, the page crashed

mursal-ashraf commented:
I figured out why `useNavigate` isn't working. It needs to be part of a Router context. As seen from the following error message in the console:
<img width="599" alt="image" src="https://github.com/mursal-ashraf/bankan/assets/89443356/01360e7c-eca0-4e6d-b25a-0d88808bb5f6">

But our `Navbar` component is not part of a Router component:
<img width="396" alt="image" src="https://github.com/mursal-ashraf/bankan/assets/89443356/f767e359-3deb-42d5-8734-3f89b682693b">

You can leave it as it is for now (without `useNavigate`), we'll make a refactoring card later that will handle this.

mursal-ashraf commented:
Can you also remove the `navItems` from App.tsx as we have removed the prop for now.

Eeyore6 committed:
Small update

Add error funciton to useLogOut.ts
Change the isLoading to true in the useSignUp and useLogin
Delete one '!' in the AuthModal in Line 45
Remove navItems in the App.tsx

mursal-ashraf commented on frontend/src/hooks/useLogOut.ts:
```diff
@@ -23,6 +23,8 @@ export const useLogOut: LogOutHook = () => {
       setResult({ ...result, isLoading: true });
       await client.auth.signOut();
```
Think you forgot to remove this one!

Eeyore6 replied:
Done.

mursal-ashraf reviewed (approved):
Good to merge with one change

Eeyore6 committed:
Update useLogOut.ts

Remove the overlapped function inside the performLogOut.ts

Eeyore6 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create and implement the Navbar](https://trello.com/c/z4Wey36o/43-create-and-implement-the-navbar)
